CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+        # owned by the current module which means that we need to call a
+        # free function of some form. For now that function is called
+        # "witx_free" and takes pointer/size/alignment arguments
+        if direction == "import":
+          defer module.witx_free(ptr, len * sizeof(direction, $ty), align(direction, $ty))
+
+        if $ty == char:
+          # note that this can skip validation due to the canonical ABI
+          read_utf8_string(ptr, len)
+        else:
+          # note that this can be optimized to a `memcpy` in some cases, and
+          # otherwise `read_from_memory` is defined below.
+          result = []
+          for i in 0..len:
+              read_from_memory(direction, $ty, ptr)
+              ptr += size(direction, $ty)
+          result
+    }
+
+    record $fields => {
+        # If this record is represented with bitflags then we pop the
+        # singular value and wrap it up in the language's representation of
+        # what bitflags are. Note that the bitflags bits are assumed valid
+        # here.
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              values.assert_next(i64) as $bitflags<$fields>
+          else:
+              values.assert_next(i32) as $bitflags<$fields>
+
+        # Otherwise if not bitflags each field is decoded individually
+        else:
+          fields = $fields.map(|ty| lift(direction, ty, values))
+          fields as $language_struct
+    }
+
+    variant $cases => {
+        # Note that this is assumed valid
+        discrim = values.assert_next(i32)
+
+        # Note that `values` would be configured here so that it
+        # automatically converts between the actual type to the desired type
+        # loaded. For example `assert_next(i32)` when the parameter is
+        # actually an i64 would auto-truncate. This is to handle the `unify`
+        # function above.
+        payload = lift(direction, $cases[discrim], values)
+
+        (discrim, payload) as $language_enum
+    }
+
+    shared-handle $resource => {
+      value = values.assert_next(i32)
+
+      # If we're importing a function then due to the acyclic nature of
+      # interface types it couldn't refer to types defined in our own
+      # module. This means that the shared handle refers to something owned
+      # by a foreign module, which means we wrap it up here.
+      #
+      # Note that it's the responsibility of the current language, in this
+      # case, to "drop" the handle. The current module is being granted
+      # access to a value of type $resource, but the current module needs to
+      # inform the outer environment when it no longer needs the resource.
+      #
+      # In the canonical ABI this means that the current module will
+      # call the `{$resource}_drop` imported (e.g.  # "fd_drop" for WASI).
+      if direction == "import":
+        assert(!$resource.owned_by_current_module);
+        value as $foreign_handle<$resource>
```

Comment:
I updated the previous section to avoid the use of `defer` to just call it at the end of the loop, and I added some wording here about how the returned value here is expected to have a destructor but otherwise I don't think `defer` is right in this location since the destructor is tied to the lifetime of the `$foreign_handle<$resource>` value.

I may be getting too confusing with all the made-up syntax and such used throughout here...